### ISSUE TITLE
Components: moving away from Emotion

### ISSUE THREAD

**Issue by ciampo** [WordPress/gutenberg#66806]
Related to #61232

## Context

The `@wordpress/components` and the `@wordpress/block-editor` packages use the [Emotion](https://emotion.sh/) for authoring styles.

Emotion is a CSS-in-JS library that dynamically computes and applies styles at runtime, allowing a developer to change the styles based on JS variables / react props. It also allows for locally scoped files (obfuscating classnames), and colocation of styles and markup.

Unfortunately, Emotion also has a few drawbacks:
- its runtime increases bundle size and slows down code execution in the browser;
- it pollutes the developer tools;
- it can conflict with other dependencies if they don't use the same Emotion version as the main project (specifically with Gutenberg, this happens when loading components in Storybook);
- their caching/serialization can lead to complicated setups with, for example, rendering components inside an iframe

### 0. Add docs and guardrails

- [x] Update `packages/components/CONTRIBUTING.md` to recommend SCSS Modules instead of Emotion for new component styles. ([#79442](https://github.com/WordPress/gutenberg/pull/79442))
- [x] Add `restricted-import` coverage to prevent new `@emotion/*` usage in `packages/components/src`, with temporary allowlists for unmigrated files. ([#79442](https://github.com/WordPress/gutenberg/pull/79442))
- [x] Keep `@wordpress/no-non-module-stylesheet-imports` enforced. ([#79442](https://github.com/WordPress/gutenberg/pull/79442))

### 1. Pilot strategy

Start migrating components from the following categories:

- [x] Easiest components to prove the general feasibility of the migration:
  - [x] `Divider` ([#79444](https://github.com/WordPress/gutenberg/pull/79444))
  - [x] `Surface` ([#79445](https://github.com/WordPress/gutenberg/pull/79445))
  - [x] `Truncate` ([#79446](https://github.com/WordPress/gutenberg/pull/79446))
- [ ] Components that validate polymorphic `as` prop compatibility:
  - [x] `View`: validates replacing Emotion's central `styled.div` polymorphic `as` behavior. ([#79443](https://github.com/WordPress/gutenberg/pull/79443))
  - [ ] `Text`: validates a representative component composed from `View`.
  - [x] `Spacer`: validates a representative component composed from `View`. ([#79449](https://github.com/WordPress/gutenberg/pull/79449))
  - [x] `Flex`: validates a representative component composed from `View`. ([#79450](https://github.com/WordPress/gutenberg/pull/79450))
  - [x] `HStack`: validates a representative component composed from `View`. https://github.com/WordPress/gutenberg/pull/79450
  - [x] `VStack`: validates a representative component composed from `View`. https://github.com/WordPress/gutenberg/pull/79450
  - [ ] `BaseControl.VisualLabel`: validates a representative styled component that currently uses Emotion's native `as` support without going through `View`.
  - [ ] `StyledLabel`: validates a representative styled component that currently uses Emotion's native `as` support without going through `View`. https://github.com/WordPress/gutenberg/pull/80001
  - [ ] `LabelText`: validates a representative styled component that currently uses Emotion's native `as` support without going through `View`.
  - [ ] Input control label wrappers: validate representative styled components that currently use Emotion's native `as` support without going through `View`.
- [ ] Risk components that test the hardest parts of the migration:
  - [x] `Theme`: validates dynamic CSS variable generation without Emotion. ([#79447](https://github.com/WordPress/gutenberg/pull/79447))
  - [x] `Spacer`: validates prop-driven layout styles and RTL replacement. ([#79449](https://github.com/WordPress/gutenberg/pull/79449))
  - [x] `Flex`: validates prop-driven layout styles and RTL replacement. ([#79450](https://github.com/WordPress/gutenberg/pull/79450))
  - [ ] `Popover`: validates portal and cross-document behavior where `StyleProvider` and `style-runtime` matter.
  - [ ] `Modal`: validates portal and cross-document behavior where `StyleProvider` and `style-runtime` matter.
  - [ ] `SlotFill`: validates portal and cross-document behavior where `StyleProvider` and `style-runtime` matter.
- [ ] Use the pilot phase to gather more information about the migration requirements, especially around the existing SCSS Modules pipeline, polymorphic `as` compatibility, and the `StyleProvider` component.

### 2. Migrate all components

Potential order:

- [ ] Start with static primitives and utilities.
- [x] Move to `View` and polymorphic compatibility work before migrating components that depend on `as`.
- [ ] Move to `View`-composed primitives and layout components once the `as` compatibility layer is proven.
- [ ] Move to layout primitives and CSS-variable-friendly components.
- [ ] Then migrate controls and form components with many variants.
- [ ] Leave overlays, menus, date/time, navigator, tools panel, and other complex flows for later PRs.

### 3. Clean up

- [ ] Simplify `StyleProvider` once Emotion cache behavior is no longer needed.
- [ ] Remove `useCx` once no component depends on Emotion `SerializedStyles`.
- [ ] Replace the Emotion-based `rtl` utility with SCSS/logical-property patterns.
- [ ] Remove Emotion Babel/plugin wiring, Storybook config, Jest serializer setup, package dependencies, and lockfile entries only after source usage is gone.
- [ ] Update any docs mentioning Emotion.

### 4. Success Criteria

The migration is complete when:

- [ ] `packages/components/src` has no active `@emotion/*` imports.
- [ ] Component styles use SCSS Modules or existing package SCSS patterns.
- [ ] Legacy public class names remain compatible.
- [ ] `View` and components using `View` preserve polymorphic `as` behavior.
- [ ] Styled components that previously relied on Emotion's native `as` support have equivalent non-Emotion behavior.
- [ ] Ref forwarding and prop passthrough continue to work for intrinsic and custom `as` targets.
- [ ] Iframe, portal, RTL, and Storybook scenarios continue to work.
- [ ] `useCx`, Emotion `rtl`, and Emotion cache behavior are no longer needed internally.
- [ ] Emotion dependencies can be removed from `@wordpress/components`.

### 5. Potential follow-up

- [ ] No `wp-build` work is currently required for this migration. We can rely on the existing CSS Modules handling.
- [ ] Open a follow-up issue only if the migration pilot reveals a concrete blocker or significant downside in the existing CSS Modules / `@wordpress/style-runtime` delivery path.

### Components and files that need migration away from Emotion

<details>
<summary>Current `packages/components/src` Emotion usage inventory</summary>

Generated from active `@emotion/*` imports in `packages/components/src`, excluding snapshots, README files, and changelogs. This intentionally lists every matching source file, so component areas may appear more than once.

- [ ] `BaseControl` — `packages/components/src/base-control/styles/base-control-styles.ts`
- [ ] `BorderBoxControl` — `packages/components/src/border-box-control/styles.ts` https://github.com/WordPress/gutenberg/pull/80437
- [ ] `BorderControl` — `packages/components/src/border-control/styles.ts` https://github.com/WordPress/gutenberg/pull/80437
- [ ] `BoxControlIcon` — `packages/components/src/box-control/styles/box-control-icon-styles.ts`
- [ ] `BoxControl` — `packages/components/src/box-control/styles/box-control-styles.ts`
- [ ] `Card` — `packages/components/src/card/card/component.tsx`
- [ ] `Card` — `packages/components/src/card/get-padding-by-size.ts`
- [ ] `Card` — `packages/components/src/card/styles.ts`
- [ ] `ColorPalette` — `packages/components/src/color-palette/styles.ts` https://github.com/WordPress/gutenberg/pull/80473
- [ ] `ColorPicker` — `packages/components/src/color-picker/styles.ts`
- [ ] `ComboboxControl` — `packages/components/src/combobox-control/styles.ts` https://github.com/WordPress/gutenberg/pull/80471
- [ ] `ConfirmDialog` — `packages/components/src/confirm-dialog/styles.ts`
- [ ] `ContextSystemProvider` test helper — `packages/components/src/context/test/context-system-provider.js`
- [ ] `CustomGradientPicker` — `packages/components/src/custom-gradient-picker/styles/custom-gradient-picker-styles.tsx`
- [ ] `CustomSelectControlV2` — `packages/components/src/custom-select-control-v2/styles.ts`
- [ ] `DatePicker` — `packages/components/src/date-time/date-picker/styles.ts`
- [ ] `DateTime` — `packages/components/src/date-time/date-time/styles.ts`
- [ ] `TimePicker` — `packages/components/src/date-time/time-picker/styles.ts`
- [x] `Disabled` — `packages/components/src/disabled/styles/disabled-styles.tsx` ([#80643](https://github.com/WordPress/gutenberg/pull/80643))
- [ ] `Divider` — `packages/components/src/divider/styles.ts` ([#79444](https://github.com/WordPress/gutenberg/pull/79444))
- [x] `Dropdown` — `packages/components/src/dropdown/styles.ts` ([#81522](https://github.com/WordPress/gutenberg/pull/81522))
- [ ] `Elevation` — `packages/components/src/elevation/hook.ts`
- [ ] `Elevation` — `packages/components/src/elevation/styles.ts`
- [ ] `FlexItem` — `packages/components/src/flex/flex-item/hook.ts` ([#79450](https://github.com/WordPress/gutenberg/pull/79450))
- [ ] `Flex` — `packages/components/src/flex/flex/hook.ts` ([#79450](https://github.com/WordPress/gutenberg/pull/79450))
- [ ] `Flex` — `packages/components/src/flex/styles.ts` ([#79450](https://github.com/WordPress/gutenberg/pull/79450))
- [ ] `FocalPointPicker` — `packages/components/src/focal-point-picker/styles/focal-point-picker-style.ts`
- [ ] `FocalPoint` — `packages/components/src/focal-point-picker/styles/focal-point-style.ts`
- [ ] `FontSizePicker` — `packages/components/src/font-size-picker/styles.ts`
- [ ] `FormTokenField` — `packages/components/src/form-token-field/styles.ts` https://github.com/WordPress/gutenberg/pull/80472
- [ ] `Grid` — `packages/components/src/grid/hook.ts`
- [ ] `InputControl` — `packages/components/src/input-control/styles/input-control-styles.tsx`
- [x] `ItemGroup` — `packages/components/src/item-group/styles.ts` ([#80797](https://github.com/WordPress/gutenberg/pull/80797))
- [ ] `Menu` stories — `packages/components/src/menu/stories/index.story.tsx`
- [ ] `Menu` — `packages/components/src/menu/styles.ts`
- [ ] `Navigator` — `packages/components/src/navigator/styles.ts`
- [ ] `NumberControl` — `packages/components/src/number-control/styles/number-control-styles.ts`
- [ ] `PaletteEdit` — `packages/components/src/palette-edit/styles.ts`
- [ ] `ProgressBar` — `packages/components/src/progress-bar/styles.ts`
- [ ] `RangeControl` — `packages/components/src/range-control/styles/range-control-styles.ts`
- [ ] `ResizableBoxResizeTooltip` — `packages/components/src/resizable-box/resize-tooltip/styles/resize-tooltip.styles.ts`
- [ ] `Scrollable` — `packages/components/src/scrollable/styles.ts`
- [ ] `SearchControl` — `packages/components/src/search-control/styles.ts` https://github.com/WordPress/gutenberg/pull/80474
- [ ] `SelectControl` — `packages/components/src/select-control/styles/select-control-styles.ts`
- [ ] `Spacer` — `packages/components/src/spacer/hook.ts` ([#79449](https://github.com/WordPress/gutenberg/pull/79449))
- [ ] `Spinner` — `packages/components/src/spinner/styles.ts`
- [ ] `StyleProvider` — `packages/components/src/style-provider/index.tsx`
- [ ] `Surface` — `packages/components/src/surface/styles.ts` ([#79445](https://github.com/WordPress/gutenberg/pull/79445))
- [ ] `Tabs` — `packages/components/src/tabs/styles.ts`
- [ ] `Text` — `packages/components/src/text/hook.ts`
- [ ] `Text` — `packages/components/src/text/styles.ts`
- [ ] `TextareaControl` — `packages/components/src/textarea-control/styles/textarea-control-styles.ts`
- [ ] `Theme` — `packages/components/src/theme/styles.ts` ([#79447](https://github.com/WordPress/gutenberg/pull/79447))
- [ ] `ToggleGroupControlOptionBase` — `packages/components/src/toggle-group-control/toggle-group-control-option-base/styles.ts`
- [ ] `ToggleGroupControl` — `packages/components/src/toggle-group-control/toggle-group-control/styles.ts`
- [ ] `ToolsPanel` stories — `packages/components/src/tools-panel/stories/index.story.tsx` https://github.com/WordPress/gutenberg/pull/80445
- [ ] `ToolsPanel` — `packages/components/src/tools-panel/styles.ts` https://github.com/WordPress/gutenberg/pull/80445
- [ ] `Truncate` — `packages/components/src/truncate/hook.ts` ([#79446](https://github.com/WordPress/gutenberg/pull/79446))
- [ ] `Truncate` — `packages/components/src/truncate/styles.ts` ([#79446](https://github.com/WordPress/gutenberg/pull/79446))
- [ ] `UnitControl` — `packages/components/src/unit-control/styles/unit-control-styles.ts`
- [ ] `BaseLabel` utility — `packages/components/src/utils/base-label.ts`
- [ ] `BoxSizing` utility — `packages/components/src/utils/box-sizing.ts`
- [ ] `Emotion` type augmentation — `packages/components/src/utils/hooks/emotion.d.ts`
- [ ] `useCx` tests — `packages/components/src/utils/hooks/test/use-cx.js`
- [ ] `useCx` — `packages/components/src/utils/hooks/use-cx.ts`
- [ ] `rtl` utility — `packages/components/src/utils/rtl.js`
- [x] `View` — `packages/components/src/view/component.tsx` ([#79443](https://github.com/WordPress/gutenberg/pull/79443))
- [ ] `ZStack` — `packages/components/src/z-stack/styles.ts`

</details>

### Additional context

- [ ] #76744 is related and potentially helpful for this migration, but not a prerequisite.
- [ ] Emotion provides native support for the `as` prop in styled components. This is used centrally by `View`, and also directly by some styled components outside `View`. The migration needs a reusable polymorphic compatibility pattern, not just a one-off `View` replacement.

### Migration guidelines and lessons learned

<details>

<summary>Click to expand</summary>

These migrations should be treated as behavior-preserving refactors. The target is the same rendered DOM, public API, and resulting styles—not an opportunity for unrelated cleanup or redesign.

### Authoring checklist

- Preserve the component’s complete public contract:
  - Keep legacy public class names, even when the new styles no longer use them.
  - Preserve the rendered element, `className`, `style`, `ref`, polymorphic rendering, prop forwarding, and supported HTML/SVG elements.
  - Avoid adding data attributes solely as styling hooks. Data attributes become observable public API; prefer private CSS Module classes applied with `clsx`.
- Use kebab-case for CSS class names.
- Preserve selector specificity and cascade behavior:
  - CSS Module scoping does not make specificity irrelevant.
  - Use `:where()` when an attribute or pseudo-class should not increase specificity. See the [Divider follow-up](https://github.com/WordPress/gutenberg/pull/79534).
  - Preserve deliberate specificity increases such as Emotion’s `&&`; for example, with a repeated selector like `.wrapper.wrapper`.
  - Check both the component and consumers that override its styles.
- Audit Emotion style composition carefully:
  - If separately composed Emotion fragments set overlapping properties, shorthand/longhand declarations, or matching nested selectors, their behavior may depend on source order.
  - In that case, combine the fragments within one `css()` call before passing the result to `cx()`. See [#79443](https://github.com/WordPress/gutenberg/pull/79443).
  - Audit downstream consumers too. The initial `View` migration audit missed a `BorderBoxControl` regression later fixed in [#79967](https://github.com/WordPress/gutenberg/pull/79967).
  - Do not mechanically combine unrelated fragments whose behavior does not depend on source order.
- Move static declarations to SCSS and represent genuinely dynamic values with CSS custom properties:
  - Preserve the existing precedence between component-generated values and a consumer’s `style` prop.
  - Check shorthand/longhand fallbacks.
  - Reset inherited internal custom properties where nested components must not inherit them.
- Do not replace physical properties with logical properties unless they are demonstrably equivalent in LTR, RTL, and relevant writing modes. For example, `padding: 0` is not equivalent to `padding-inline: 0`.
- Prefer WPDS tokens when they produce the same result. Treat intentional token or visual changes as separate, documented changes rather than hiding them inside the migration.
- Preserve necessary browser-specific declarations, such as `-webkit-box-orient`, including narrowly scoped lint suppressions where required.
- After removing Emotion code, check for:
  - Obsolete Emotion lint or type suppressions.
  - Unused style utilities, constants, and variables, including values under `packages/components/src/utils/`.
  - Unused imports and dependencies.
  - Avoid bundling unrelated cleanup into the migration.
- For components rendered through portals, SlotFill, or another document:
  - Verify that the generated CSS is registered in the target document.
  - Do not remove `StyleProvider` or related style-registration behavior merely because Emotion is no longer used by the migrated component.
  - Test against a real iframe document and assert computed styles, as demonstrated by [#80384](https://github.com/WordPress/gutenberg/pull/80384).

### Verification checklist

- Test the default element and every supported polymorphic path.
- Test refs, prop forwarding/filtering, `className`, `style`, and legacy public class names.
- Test conditional variants and combinations of dynamic values.
- Inspect computed styles and matched declarations when debugging:
  - Compare specificity.
  - Compare stylesheet source order.
  - Check shorthand/longhand interactions.
  - Check whether consumer overrides still win.
  - Check whether CSS custom properties are inherited or overwritten unexpectedly.
- Review relevant Storybook examples and downstream consumers, not only the migrated component in isolation.
- Test RTL when directional styles are involved.
- Test portal/iframe rendering when styles may cross document boundaries.
- Generate snapshots for the **whole repository** with `npm run test:unit:update`.
  - Review every snapshot change rather than assuming CSS Module class-name changes are harmless.
  - Unexpected changes can reveal altered DOM output, class composition, prop forwarding, or downstream regressions.
- Run focused tests and linting first, followed by the whole-repository gates:
  - `npm run test:unit -- <test-path> --runInBand`
  - `npm run lint:js -- <changed-files>`
  - `npm run lint:css -- <changed-styles>`
  - `npm run build`
  - `npm run test:unit:update`
  - `npm run other:check-local-changes`
  - `git diff --check`
- Re-review the final diff after the last selector or composition change. Earlier approvals may no longer apply after a specificity-sensitive edit.
- If a broad test fails, rerun the exact file or shard before attributing it to the migration.

The central review question should be: **does this produce the same observable component and the same winning CSS declarations for both the component and its consumers?**
</details>

cc @WordPress/gutenberg-components

**Comment by ciampo** [comment]
Update:

Folks from @WordPress/gutenberg-components spent some time in the past couple of days testing a few approaches.

### Long-term goal

As the long-term solution, we discussed replacing Emotion with [CSS modules](https://github.com/css-modules/css-modules).

@mirka started experimenting with adding the necessary steps to the `build:packages` npm script, with initially encouraging results.

Goals:

- [ ] create a working build step, which creates the obfuscated class names mapping and appends the styles to the package's existing CSS bundle (both LTR and RTL)
- [ ] refactor each component from Emotion to CSS modules
  - [ ] replaced `styled` components with the underlying un-styled component and move styles to classname
  - [ ] replace dynamic props / variables with other equivalent logic (HTML attributes / CSS variables / different class names)
  - [ ] find a way around using "components as selectors" 
- [ ] Remove `StyleProvider` component and its usages across the codebase, `useCx` hook, `rtl` utility
- [ ] remove any remaining usage of `@emotion/*` dependencies (storybook, tests setup, etc)

One thing to consider, is whether we should use "SCSS" modules instead, so that all styles are authored in SCSS across the whole repository. Using SCSS will also allow to re-use SCSS variables, without the need for re-creating them in another language (as we currently do with JS variables for CSS-in-JS styles).

### Intermediate step

We discussed applying an intermediate step, where we'd keep the CSS-in-JS code as "untouched" and possible, and we'd focus on swapping Emotion with another (no-runtime) CSS-in-JS. We looked at [linaria](https://linaria.dev/), [StyleX](https://stylexjs.com/), and [vanilla-extract](https://vanilla-extract.style/).

- Linaria and vanilla-extract only offer plugins for known build tools (ie. webpack, rollup, vite..). Unfortunately, this is not very compatible with the custom way Gutenberg packages are build (ie. a node script applying a babel transformation on each individual JS/TS/TSX file — no bundling — and a compilation of any SCSS styles into one bundled CSS file). @ciampo experimented with this part, but didn't get very far: using any build tool brings in additional overhead and new issues, and otherwise using the private APIs of linaria and vanilla-extract is not straightforward
- StyleX does offer a babel plugin, but it will still require custom code to extract the styles and write them to disk, and to correctly watch file changes;
- In any case, many Emotion features are not compatible with these libraries — in particular, any "dynamic" variable (or any variable that is imported from another file) may not work as expected, meaning that we'd need to refactor the CSS-in-JS code anyway.

For the above reasons, @ciampo 's initial conclusion is that it may be a better strategy to skip the intermediate step, and instead refactor the Emotion code directly to whatever long-term solution we want to adopt (eg. CSS modules).

**Comment by mirka** [comment]
A potential complication that I'm not seeing mentioned is the `as` prop, and that many components are composed from the `View` component (these two problems are related, since `View` is often used to provide the `as` functionality). We'll need some kind of compatibility layer for the `as` handling.

> Evaluate whether (S)CSS Module output from package builds should be extracted into `build-style/style.css` and `style-rtl.css` instead of being delivered through `@wordpress/style-runtime` ([@mirka](https://github.com/mirka) I'm mentioning this because of the "
> create a working build step, which creates the obfuscated class names mapping and appends the styles to the package's existing CSS bundle (both LTR and RTL)" part in the previous message, although I'm not sure this is still relevant?)

No, not relevant. We can just rely on the existing CSS module handling.

**Comment by ciampo** [comment]
@mirka good point re. the `as` prop, I had forgotten about it 😅 

My initial idea is to create an internal `PolymorphicElement` utility component that takes care of the main compat parts (ie ref forwarding, prop passthrough, and choosing the final element), which can be used for both `View` and in all other circumstances where the `as` prop support was provided directly by Emotion's `styled` APIs.

I've updated the plan to mention the `as` prop explicitly, and updated the issue's original post too.

**Comment by im3dabasia** [comment]
Hey @ciampo and @mirka,

I tried migrating the `StyledLabel` component since it is an independent component. Here's the PR: https://github.com/WordPress/gutenberg/pull/80001.

I'd really appreciate your feedback on how we can improve it and get it ready to merge into trunk.

**Comment by ciampo** [comment]
Given the high number of PRs opened (and their average low quality which results in a lot of time spent reviewing those PRs), I'm going to assign this issue to myself and iterate at a later point, as it's more time effective
